File: java/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/DensityHitList.java
package vsdk.toolkit.raycasting.bidirectionalRaytracing;

import vsdk.toolkit.common.Error;

public class DensityHitList {
    protected static final int DHL_ARRAY_SIZE = 20;
    protected DensityHitArray first;
    protected DensityHitArray last;
    protected int numHits;
    protected int cacheLowerLimit;
    protected DensityHitArray cacheCurrent;

    public DensityHitList() {
        first = new DensityHitArray(DHL_ARRAY_SIZE);
        last = first;
        cacheCurrent = null;
        numHits = 0;
        cacheLowerLimit = 0;
    }

    public void add(DensityHit hit) {
        if ( !last.add(hit) ) {
            // New array needed

            last.setNext(new DensityHitArray(DHL_ARRAY_SIZE));
            last = last.getNext();

            last.add(hit); // Supposed not to fail
        }

        numHits++;
    }

    public int storedHits() {
        return numHits;
    }

    public DensityHit get(int i) {
        if ( i >= numHits ) {
            Error.fatal(-1, "DensityHitList::operator[]", "Index 'i' out of getBoundingBox");
        }

        if ( cacheCurrent == null || (i < cacheLowerLimit) ) {
            cacheCurrent = first;
            cacheLowerLimit = 0;
        }

        // Wanted point is beyond cacheCurrent
        while ( i >= cacheLowerLimit + DHL_ARRAY_SIZE ) {
            cacheCurrent = cacheCurrent.getNext();
            cacheLowerLimit += DHL_ARRAY_SIZE;
        }

        // Wanted point is in current cache block

        return cacheCurrent.get(i - cacheLowerLimit);
    }
}

File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/DensityHitList.ts
import { Error as VsdkError } from "../../common/Error";
import { DensityHit } from "./DensityHit";
import { DensityHitArray } from "./DensityHitArray";

export class DensityHitList {
  protected static readonly DHL_ARRAY_SIZE = 20;
  protected first: DensityHitArray;
  protected last: DensityHitArray;
  protected numHits: number;
  protected cacheLowerLimit: number;
  protected cacheCurrent: DensityHitArray | null;

  public constructor() {
    this.first = new DensityHitArray(DensityHitList.DHL_ARRAY_SIZE);
    this.last = this.first;
    this.cacheCurrent = null;
    this.numHits = 0;
    this.cacheLowerLimit = 0;
  }

  public add(hit: DensityHit): void {
    if (!this.last.add(hit)) {
      this.last.setNext(new DensityHitArray(DensityHitList.DHL_ARRAY_SIZE));
      this.last = this.last.getNext() as DensityHitArray;
      this.last.add(hit);
    }

    this.numHits++;
  }

  public storedHits(): number {
    return this.numHits;
  }

  public get(i: number): DensityHit {
    if (i >= this.numHits) {
      VsdkError.fatal(-1, "DensityHitList::operator[]", "Index 'i' out of getBoundingBox");
    }

    if (this.cacheCurrent === null || i < this.cacheLowerLimit) {
      this.cacheCurrent = this.first;
      this.cacheLowerLimit = 0;
    }

    while (i >= this.cacheLowerLimit + DensityHitList.DHL_ARRAY_SIZE) {
      this.cacheCurrent = this.cacheCurrent!.getNext();
      this.cacheLowerLimit += DensityHitList.DHL_ARRAY_SIZE;
    }

    return this.cacheCurrent!.get(i - this.cacheLowerLimit);
  }
}
